lib/libutils/ext/include/compiler.h

```diff
@@ -246,4 +249,25 @@
 #define __compiler_atomic_store(p, val) \
 	__atomic_store_n((p), (val), __ATOMIC_RELAXED)
 
+#ifdef __HAVE_BUILTIN_SPECULATION_SAFE_VALUE
+#define __compiler_speculation_safe_value(v) __builtin_speculation_safe_value(v)
+#else
+
+#if defined(ARM32) || defined(ARM64)
+/*
+ * With compiler support this can be done more optimized, but as a fallback
+ * these instructions:
+ * isb
+ * dsb sy
+ * guarantees ([1]) that there will be no speculation crossing this point.
+ * Link [1]: https://github.com/ARM-software/arm-trusted-firmware/commit/9edd89127928708b9f4dd36bd3c5d16b62d27cb5
```

Comment:
According to the comment in the linked commit, you also need `-mtrack-speculation` for better performance of `__builtin_speculation_safe_value()`, otherwise the compiler will fallback to `isb; dsb sy;`. So I think this should be detected and added at compile time, in `mk/compile.mk` maybe (I'm thinking Linux's `cc-option` function). Could be done in a separate commit.

Reply:
@jforissier, thanks, please propose something.

Reply:
@jenswi-linaro please see https://github.com/OP-TEE/optee_os/pull/3486.

If you wish you may pick up this patch before "core: protect syscall table lookup from speculation", in which case you may add:

```diff
diff --git a/core/arch/arm/tee/sub.mk b/core/arch/arm/tee/sub.mk
index e8b07ee4..4ff085fd 100644
--- a/core/arch/arm/tee/sub.mk
+++ b/core/arch/arm/tee/sub.mk
@@ -3,6 +3,7 @@ srcs-$(CFG_ARM32_core) += arch_svc_a32.S
 srcs-$(CFG_ARM64_core) += arch_svc_a64.S
 srcs-$(CFG_CACHE_API) += svc_cache.c
 srcs-y += arch_svc.c
+cflags-arch_svc.c-y += $(call cc-option,-mtrack-speculation)
 else
 srcs-y += svc_dummy.c
 endif
```

Reply:
OK, I'll wait for #3486 to be merged and then I'll rebase and add this patch.

Reply:
Sounds good.

Reply:
Dependent PR merged.

Reply:
Added